src/designspace.rs

```diff
@@ -0,0 +1,213 @@
+//! Reading and writing designspace files.
+
+#![deny(rustdoc::broken_intra_doc_links)]
+
+use quick_xml::de::from_reader;
+use std::{fs::File, io::BufReader, path::Path};
+
+use crate::error::DesignSpaceLoadError;
+
+/// A [designspace]].
+///
+/// [designspace]: https://fonttools.readthedocs.io/en/latest/designspaceLib/index.html
+#[derive(Clone, Debug, Default, PartialEq, Deserialize)]
+#[serde(rename = "designspace")]
+pub struct DesignSpaceDocument {
+    /// Design space format version.
+    pub format: f32,
+    /// One or more axes.
+    pub axes: Axes,
+    /// One or more sources.
+    pub sources: Sources,
+    /// One or more instances.
+    pub instances: Instances,
+}
+
+/// https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#axes-element
+#[derive(Clone, Debug, Default, PartialEq, Deserialize)]
+#[serde(rename = "axes")]
+pub struct Axes {
+    /// One or more axis definitions.
+    pub axis: Vec<Axis>,
+}
+
+/// A [axis]].
+///
+/// [axis]: https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#axis-element
+#[derive(Clone, Debug, Default, PartialEq, Deserialize)]
+#[serde(rename = "axis")]
+pub struct Axis {
+    /// Name of the axis that is used in the location elements.
+    pub name: String,
+    /// 4 letters. Some axis tags are registered in the OpenType Specification.
+    pub tag: String,
+    /// The default value for this axis, in user space coordinates.
+    pub default: f32,
+    /// Records whether this axis needs to be hidden in interfaces.
+    #[serde(default)]
+    pub hidden: bool,
+    /// The minimum value for a continuous axis, in user space coordinates.
+    pub minimum: Option<f32>,
+    /// The maximum value for a continuous axis, in user space coordinates.
+    pub maximum: Option<f32>,
+    /// The possible values for a discrete axis, in user space coordinates.
+    pub values: Option<Vec<f32>>,
+    /// Mapping between user space coordinates and design space coordinates.
+    pub map: Option<Vec<AxisMapping>>,
+}
+
+/// Maps one input value (user space coord) to one output value (design space coord).
+#[derive(Clone, Debug, Default, PartialEq, Deserialize)]
+#[serde(rename = "map")]
+pub struct AxisMapping {
+    /// user space coordinate
+    pub input: f32,
+    /// designspace coordinate
+    pub output: f32,
+}
+
+/// https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#sources-element
+#[derive(Clone, Debug, Default, PartialEq, Deserialize)]
+#[serde(rename = "sources")]
+pub struct Sources {
+    /// One or more sources.
+    pub source: Vec<Source>,
+}
```

Comment:
ditto, would be nice to avoid the wrapper type.

Reply:
Done